Ladder step2

AppleCEO opened the pull request:
Step4-2 요구사항 반영

AppleCEO committed:
Ladder step4-1 input, process

AppleCEO committed:
Step4-1

AppleCEO committed:
Unused function delete

AppleCEO committed:
If let devide per variable

AppleCEO committed:
Variable name changed

AppleCEO committed:
Enum type name first character Upper cased

AppleCEO committed:
Function output changed to tuple

AppleCEO committed:
Variable name changed to simple and excatly

AppleCEO committed:
Input to repaet

AppleCEO committed:
Index variable name changed

AppleCEO committed:
Function ladderPrint variable name changed

AppleCEO committed:
Function comment add

AppleCEO committed:
Check variable changed, delete forced unwrapping

AppleCEO committed:
Human number changed

AppleCEO committed:
Change made to make

AppleCEO committed:
Merge branch 'kiljh-me' of https://github.com/kiljh-me/swift-laddergame into kiljh-me

AppleCEO committed:
Step4-1 에서 지적 받은 사항 수정

input 함수에서 print 등 다른 작업을 하지 말기
입력받은 String? 값을 강제로 Int 형변환 시키는 부분이 있었는데 ?? 0 으로 기본값을 설정해주고 nil일 경우 죽지않도록 수정

AppleCEO committed:
요구사항 2 적용

사다리가 옆으로 연속해서 |-|-| 나오지 않도록 검증한다.
바로 왼쪽 사다리가 이어져있으면 무조건 공백으로 들어가도록 구성

AppleCEO committed:
요구사항 3. 메인 함수에 반영

while문 안에서 입력한 숫자를 if문으로 확인하던 것을 do while의 형태로 while문 위에서 먼저 입력값을 받고 입력값이 부적합하다면 while문을 돌면서 입력하도록 수정

AppleCEO committed:
요구사항 3 반영 ladderMake에서 if 제거, 함수로 처리

사다리를 만들고 저장하는 함수 ladderMake 에서 전에 사다리가 연결되있는지 확인하기 위해서 if 가 사용되고 있었는데 depth를 줄이기 위해 별도 함수에서 if 등의 작업을 처리하도록 수정

AppleCEO committed:
inputFromUser 함수에서 검증하는 부분 별도 함수로 작성

10줄 이상 넘어가기 때문에 검증하는 부분 떼어내서 별도 함수로 작성

AppleCEO committed:
inputFromUser 함수에서 튜플 반환하던 것을 human, height 변수로 반환하도록 수정

input 이라고 튜플을 반환받으니까 각 요소에 대한 변수명이 없으니 어떤 데이터를 담은 튜플값인지 알 수 없기 때문에 명확하게 변수 이름으로 데이터 받아오도록 수정

godrm commented:
지금 PR은 Merge는 가능하지만, 중간에 Merge Request 가 있어서 커밋 이력이 사라지도록 저도 Merge Request로 합쳐서 머지할 수 밖에 없습니다. 그래도 될까요?
Merge Request 대신에 커밋들만 pick 해서 보내는 방법도 있습니다.

AppleCEO commented:
Merge Request 로 합쳐서 머지해주시면 될 것 같습니다. 감사합니다!

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -24,40 +24,39 @@ func inputFromUser () -> (Int, Int) {
     var height = 0
+    (human, height) = inputVerify(optionalHuman, optionalHeight)
     
+    return (human, height)
```
함수가 한 가지 책임만 갖도록 함수를 분리해보세요. 공통적으로 반복되는 부분은 일반화해보세요.
여기서는 분리한 함수를 조합해서 처리해보세요.

AppleCEO replied:
human과 height 입력받는 함수를 분리했습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -24,40 +24,39 @@ func inputFromUser () -> (Int, Int) {
-        print("참여할 사람과 최대 사다리 높이를 정확히 입력해주세요.")
-    }
-    
     return (human, height)
```
- 입력 처리 때문에 `input` 단어를 자꾸만 사용하는데, 이 함수는 어떤 역할을 하는 함수인가요?
`검증하는 함수`라고 쓰여있으면 검증과 관련된 동사가 앞에 나와야 할 것 같네요. 
- 각각 사용자수, 높이 값을 검증하는 함수를 분리해보세요.

AppleCEO replied:
사용자수와 높이를 검증하는 함수를 분리하였습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -24,40 +24,39 @@ func inputFromUser () -> (Int, Int) {
-    let human = input.0
-    let height = input.1
+func ladderMake (_ human:Int, _ height:Int) -> [[String]] {
+    var ladderMakeCheck:LadderMakeChecker
```
- 함수이름은 동사로 시작해야 합니다. 
- `ladderMakeCheck` 변수이름, `LadderMakeChecker` 타입 이름은 어색하고 어떤 의도인지 명확하지 않습니다.

AppleCEO replied:
함수이름은 동사로 시작하도록 모두 수정했습니다. ladderMakeCheck 와 LadderMakeChecker 타입은 제거하고 Bool 값으로 동작하도록 변경했습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -24,40 +24,39 @@ func inputFromUser () -> (Int, Int) {
-    let height = input.1
+func ladderMake (_ human:Int, _ height:Int) -> [[String]] {
+    var ladderMakeCheck:LadderMakeChecker
     var ladder : [[String]] = Array(repeating: Array(repeating: "",count:human ), count: height)
```
- 2차원 배열인데도 그냥 ladder 사다리로 표현하면 (복수에 무언가 들어있는) 배열이라는 것을 인지하기 어렵습니다. 
- 사다리를 `[[String]]` 문자열 기준으로 만드는 것보다는 발판이 있냐 없냐 정보만 있으면 되니까 `[[Bool]]`로 만드는게 어떨까요?
출력을 위해서 사다리를 표현하는 데이터 구조를 문자열로 만들어야 할 필요는 없습니다.

AppleCEO replied:
배열의 이름을 ladderBoolBoard 로 변경하였습니다.
사다리를 Bool 로 변경했습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -82,11 +93,14 @@ func ladderPrint (_ ladder:[[String]]) -> () {
+    (human, height) = inputFromUser()
+    while human < 1 && height < 1 {
+        print("참여할 사람과 최대 사다리 높이를 정확히 입력해주세요.")
+        (human, height) = inputFromUser()
```
`(human, height) = inputFromUser()` 두 번 나오는데 상단은 없어도 되지 않나요?

AppleCEO replied:
위에서 먼저 한번 입력받습니다. 입력한 값에 문제가 있다면 while문에 들어가서 오류를 출력하고 다시 입력을 받습니다. 그러한 이유로 두 번 호출하였습니다.

godrm replied:
그렇군요. 중복된 부분이 거슬렸던 것 같네요. 일단 while 구문에 들어가고 정상적인 경우 빠져나가는 방법도 있겠네요.

AppleCEO replied:
그러한 방법으로 먼저 구현하고자 생각했었습니다.  일단 while 구문에 들어가게 되면 if문으로 제대로 된 값인지 확인하고 break 를 해주는 부분이 있어야지 오류가 출력되지 않습니다. 그러면 depth가 늘어나기 때문에 위와 같이 구현했습니다.

godrm replied:
그렇군요. 넘어갑시다.

godrm reviewed (commented):
다음 피드백을 확인해보세요.

AppleCEO committed:
참여할 사람 숫자와 최대 사다리 높이를 입력받는 함수와 검증하는 함수를 나누어라

인풋을 하는 함수를 두가지 같이 받았는데 제이케이의 피드백을 받고 두개로 나누었습니다. 또한 검증을 하는 함수도 두개를 별도로 나누어봤습니다.

AppleCEO committed:
함수이름 동사로 시작하도록 변경

함수이름은 동사로 시작해야한다는 피드백 반영
사다리를 이을지 말지를 enum으로 작성했었는데 boolean이 합당하다고 생각해서 변경

AppleCEO committed:
2차원 배열을 String에서 Bool로 변경

어차피 사다리를 연결할지 안할지 여부만 저장되고 있기 때문에 굳이 String 배열을 사용할 필요 없이 Bool 배열로 변경

AppleCEO committed:
Bool을 붙여서 ladder라는 변수를 ladderBoolBoard 로 변경

ladder 변수에 어떤 형식이 들어있는지 모르겠다는 피드백을 보고 Bool을 추가하였으며 뒤에 Board 를 추가함으로써 2차원 배열임을 명시

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,100 +8,103 @@
-    return (human, height)
+    
+    return heightNumber
 }
```
입력받는 함수와 숫자 검증하는 함수에서 공통부분을 일반화해보세요. 비슷한 부분이 많아 보이네요.

AppleCEO replied:
공통함수로 만들었고 다른 부분만 if문으로 확인해서 실행시키도록 구성하였습니다.

AppleCEO committed:
사람 수와 사다리 높이를 입력받고 검증하는 함수에서 공통된 부분을 추려서 입력하는 함수 하나와 검증하는 함수 하나로 합침

사람 숫자를 입력하는 함수와 최대 사다리 높이를 입력하는 함수가 공통된 부분이 많다는 피드백을 받고 변수에 따라 다른 부분은 별도 처리하고 함수를 합치게 되었습니다.

AppleCEO committed:
사다리를 출력할 때 한줄을 출력하는 함수를 만들어라

사다리를 이중 for문 돌면서 출력되던 것을 depth 를 줄이기 위해서 한줄을 출력하는 함수를 별도로 만들어서 실행하도록 구현

AppleCEO committed:
사다리를 한줄씩 만드는 함수 작성

2중 for문을 줄이기 위해서 사다리를 한줄을 만드는 함수를 작성

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,61 +8,55 @@
+    switch variableName {
+    case "humanNumber" : print("참여할 사람은 몇 명 인가요?")
+    case "heightNumber" : print("최대 사다리 높이는 몇 개인가요?")
+    default: print("어떤 값을 입력하실건가요?")
```
variableName를 매개변수로 받는다면 그냥 그걸 print() 하면 어떤가요?
문자열 비교까지 해서 다른 문자열로 바꿔야 할까요?
하위 함수에서 필요한 내용을 상위 함수에서 결정해주는 것은 매우 자연스러운 것입니다.

AppleCEO replied:
문자열 비교를 하지 않고 변수명이 직접 출력되어서 입력을 받도록 수정하였습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,61 +8,55 @@
+    default: print("어떤 값을 입력하실건가요?")
+    }
+    let optionalInput = readLine()
+    let inputNumber = verifyInput(variableName, optionalInput)
```
사용자의 입력을 받는 함수라면 입력만 받고,
입력받은 문자열을 비교하고 검증하는 함수는 분리해보세요. 
- 검증 단계는 입력 단계와 처리 단계와 같은 추상화 수준이 아닐까요?

AppleCEO replied:
검증하는 함수를 제거하였습니다. 검증을 하는 함수라고 하지만 입력받은 데이터를 Int로 형변환하는 기능이 주된 기능이었습니다. 이러한 기능은 입력하는 함수 내에서 충분히 처리가능하다고 생각을 했습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,61 +8,55 @@
-    var heightNumber = 0
+/// 사다리를 만들고 저장하는 함수
+func makeLadder (_ humanNumber: Int, _ heightNumber: Int) -> [[Bool]] {
+    var ladderBoolBoard : [[Bool]] = Array(repeating: Array(repeating: false, count:humanNumber), count: heightNumber)
```
`-Number` 라는 단어는 `숫자`를 의미합니다. `개수`, `최대수`를 의도한걸까요?
사람숫자와 높이숫자가 보다 명확한 의도한 내용을 설명하는 단어는 없을까요?

AppleCEO replied:
명확한 의도를 가진 변수명으로 수정하였습니다.

AppleCEO committed:
참여하는 사람의 숫자, 최대 사다리 높이 변수명 수정

humanNumber -> numberOfPeopleToParticipate
heightNumber -> maximumLadderHeight
어떤 변수를 입력하세요. 라는 부분을 변수명이 나오도록 수정

AppleCEO committed:
검증하는 함수 제거

입력을 받는 함수에서 숫자로 변환하는 기능도 충분히 가능한 상태였다고 생각이 들어서 검증하는 함수는 제거하였다. 입력을 받는 함수에서 숫자로 변환한다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -93,16 +80,16 @@ func printLadderRow (ladderBoolRow:[Bool]) -> () {
-        humanNumber = inputFromUser(variableName: "humanNumber")
-        heightNumber = inputFromUser(variableName: "heightNumber")
+        numberOfPeopleToParticipate = inputFromUser(variableName: "numberOfPeopleToParticipate")
+        maximumLadderHeight = inputFromUser(variableName: "maximumLadderHeight")
```
이거 실행해본건가요? "numberOfPeopleToParticipate", "maximumLadderHeight" 를 출력하려는 의도인가요?

AppleCEO replied:
numberOfPeopleToParticipate : 5
maimumLadderHeight : 4 

이런식으로 입력받도록 구성했습니다.

godrm replied:
너무 불친절한 프로그램이군요 ㅎㅎ

godrm replied:
보통 업무를 할때도 입력/출력 문구는 개발팀에서 결정하지 않습니다. 유연하게 대처하도록 만드는게 개발자가 할 일입니다.

AppleCEO replied:
입력 안내 문구 변경에 유연하게 대처하도록 별도 enum 타입을 선언하였고 변수 별로 입력 안내 멘트를 출력하는 함수를 구현하였습니다.

AppleCEO committed:
변수 별 입력 안내 멘트 enum 타입 추가, 출력 함수 구현

입력 안내 문구 변경에 유연하게 대처하도록 별도 enum 타입을 선언하였고 안내 멘트를 출력하는 함수를 구현하였습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -22,6 +33,11 @@ func inputFromUser (variableName: String) -> (Int) {
 
+/// 변수 별 입력 안내 멘트 출력
+func printInputGuideMent (_ inputVariableName: String) {
+    for guide in InputGuideMent.allCases { if guide.variableName == inputVariableName { print(guide.rawValue) } }
```
이건 좀 오버 엔지니어링이네요. 그냥 `enum InputGuideMent` 타입을 사용하면 되는게 아닐까요?

AppleCEO replied:
enum InputGudieMent 타입을 사용해서 for문 없이 프린트하도록 수정하였습니다.

AppleCEO committed:
변수 별 입력 안내 멘트 출력하는 함수 수정

모든 케이스를 반복하는 부분이 오버 엔지니어링이었다. enum 타입을 사용해서 간단히 출력하도록 구성을 변경하였다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -34,8 +29,14 @@ func inputFromUser (variableName: String) -> (Int) {
+        guide = InputGuideMent.maximumLadderHeight
+    }
+    
+    print(guide.rawValue)
```
printInputGuideMent() 나 inputFromUser() 매개변수 자체를 enum 타입으로 바꾸는 건 어떨까요?
타입을 만들고 활용을 못하고 더 복잡한 방법을 쓰고 있네요. if/switch로 비교하지말고 직접 쓰는게 더 좋습니다.

AppleCEO replied:
inputFromUser 매개변수를 enum 타입으로 변경하였습니다. enum 타입의 이름이 불명확하다고 생각되어서 InputableVariableName 으로 변경해보았습니다. 피드백 부탁드립니다. 감사합니다.

AppleCEO committed:
inputFromUser() 매개변수를 enum 타입으로 변경

if swith로 비교해서 enum 타입을 사용하던 것을 직접 매개변수로 넘겨서 사용하도록 수정

AppleCEO committed:
입력할 수 있는 변수의 이름에 모음인 inputVariable enum 타입의 이름 변경

inputVairable 이라고 enum 타입을 선언하였으나 이름이 불명확하다고 생각되어서 InputableVariableName 이라고 변경하였습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -97,13 +86,13 @@ func printLadderRow (ladderBoolRow:[Bool]) -> () {
-        numberOfPeopleToParticipate = inputFromUser(variableName: "numberOfPeopleToParticipate")
-        maximumLadderHeight = inputFromUser(variableName: "maximumLadderHeight")
+        numberOfPeopleToParticipate = inputFromUser(InputableVariableName.numberOfPeopleToParticipate)
+        maximumLadderHeight = inputFromUser(InputableVariableName.maximumLadderHeight)
```
타입을 명시하고 나면 `.maximumLadderHeight`만 넣어도 될겁니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,37 +8,26 @@
-    if variableName == "numberOfPeopleToParticipate" {
+    if variableName == InputableVariableName.numberOfPeopleToParticipate {
         inputNumber -= 1
     }
```
- 여기 함수에 input 이라는 단어가 몇 번 나올까요?
입력을 처리하는 함수라고 해서 input을 자꾸만 사용할 필요는 없습니다.
- `optionalInput` 도 옵셔널을 명시할 필요는 없고, `stringInput`도 그 변수 역할이 무엇인지 의도를 표현하세요. 
- `InputableVariableName.numberOfPeopleToParticipate` 일때는 왜 -1 처리하는건가요? 
- 예외적인 상황이더라도 return 값으로 표현하지마세요. 대신 다른 방법을 고려하세요.

AppleCEO committed:
타입을 명시한 상태에서 enum 타입이름까지 작성한 부분 수정

.case 이름만으로 확인할 수 있는데 enum 타입명까지 작성해서 그 부분 제거

AppleCEO committed:
함수 inputFromUser 내에서 무의미하게 변수명 뒤에 붙던 input 제거

AppleCEO committed:
무의미하게 옵셔널로 받아서 string으로 unwrapping하고 int로 형변환하던 것을 한 줄에 처리하도록 수정

AppleCEO committed:
참여하는 사람의 숫자를 입력받고 바로 중간 사다리의 숫자로 1을 빼던 것을 제거

참여하는 사람의 숫자가 줄어드는 것은 아니기 때문에 1을 빼던 것은 정확하지 않고 >편법이었습니다. 사다리를 만들 때 하나를 빼고 만들도록 수정했습니다.

AppleCEO committed:
예외처리 구현

에러 타입인 LadderGameErrors enum type 선언
do catch, guard throw 를 통해 예외처리 구현

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -10,41 +10,57 @@ import Foundation
-func makeLadder (_ numberOfPeopleToParticipate: Int, _ maximumLadderHeight: Int) -> [[Bool]] {
-    var ladderBoolBoard : [[Bool]] = Array(repeating: Array(repeating: false, count:numberOfPeopleToParticipate), count: maximumLadderHeight)
+func makeLadder (_ numberOfParticipants: Int, _ maximumLadderHeight: Int) -> [[Bool]] {
+    var ladderBoolBoard : [[Bool]] = Array(repeating: Array(repeating: false, count:numberOfParticipants-1), count: maximumLadderHeight)
```
사소한 표현의 차이지만 `numberOfParticipants` 와 `maximumLadderHeight` 형태는 왜 다를까요?
후자처럼 `numberOf-` 가 없어도 된다면 앞에서도 뺄 수 있지 않을까요?

AppleCEO committed:
사람 숫자와 최대 사다리 높이를 저장하는 변수 이름 변경

제이케이의 피드백처럼 numberOf를 생략해보았다.
최대 사다리 높이에서 이미 함수명이나 전후 사정으로 사다리라는 것은 알 수 있으므로 maximumHeight 로 줄였다.

godrm commented:
다음 단계로 갑시다!
